File: PolicyPlus/AdmxFile.vb
Imports System.Xml
Public Class AdmxFile
    Public SourceFile As String
    Public AdmxNamespace As String
    Public SupersededAdm As String
    Public MinAdmlVersion As Decimal
    Public Prefixes As New Dictionary(Of String, String)
    Public Products As New List(Of AdmxProduct)
    Public SupportedOnDefinitions As New List(Of AdmxSupportDefinition)
    Public Categories As New List(Of AdmxCategory)
    Public Policies As New List(Of AdmxPolicy)
    Private Sub New()
    End Sub
    Public Shared Function Load(File As String) As AdmxFile
        ' ADMX documentation: https://technet.microsoft.com/en-us/library/cc772138(v=ws.10).aspx
        Dim admx As New AdmxFile
        admx.SourceFile = File
        Dim xmlDoc As New XmlDocument
        xmlDoc.Load(File)
        Dim policyDefinitions = xmlDoc.GetElementsByTagName("policyDefinitions")(0)
        For Each child As XmlNode In policyDefinitions.ChildNodes
            Select Case child.LocalName
                Case "policyNamespaces" ' Referenced namespaces and current namespace
                    For Each policyNamespace As XmlNode In child.ChildNodes
                        Dim prefix = policyNamespace.Attributes("prefix").Value
                        Dim fqNamespace = policyNamespace.Attributes("namespace").Value
                        If policyNamespace.LocalName = "target" Then admx.AdmxNamespace = fqNamespace
                        admx.Prefixes.Add(prefix, fqNamespace)
                    Next
                Case "supersededAdm" ' The ADM file that this ADMX supersedes
                    admx.SupersededAdm = child.Attributes("fileName").Value
                Case "resources" ' Minimum required version
                    admx.MinAdmlVersion = Decimal.Parse(child.Attributes("minRequiredRevision").Value, Globalization.CultureInfo.InvariantCulture)
                Case "supportedOn" ' Support definitions
                    For Each supportInfo As XmlNode In child.ChildNodes
                        If supportInfo.LocalName = "definitions" Then
                            For Each supportDef As XmlNode In supportInfo.ChildNodes
                                If supportDef.LocalName <> "definition" Then Continue For
                                Dim definition As New AdmxSupportDefinition
                                definition.ID = supportDef.Attributes("name").Value
                                definition.DisplayCode = supportDef.Attributes("displayName").Value
                                definition.Logic = AdmxSupportLogicType.Blank
                                For Each logicElement As XmlNode In supportDef.ChildNodes
                                    Dim canLoad As Boolean = True
                                    If logicElement.LocalName = "or" Then
                                        definition.Logic = AdmxSupportLogicType.AnyOf
                                    ElseIf logicElement.LocalName = "and" Then
                                        definition.Logic = AdmxSupportLogicType.AllOf
                                    Else
                                        canLoad = False
                                    End If
                                    If canLoad Then
                                        definition.Entries = New List(Of AdmxSupportEntry)
                                        For Each conditionElement As XmlNode In logicElement.ChildNodes
                                            If conditionElement.LocalName = "reference" Then
                                                Dim product = conditionElement.Attributes("ref").Value
                                                definition.Entries.Add(New AdmxSupportEntry With {.ProductID = product, .IsRange = False})
                                            ElseIf conditionElement.LocalName = "range" Then
                                                Dim entry As New AdmxSupportEntry With {.IsRange = True}
                                                entry.ProductID = conditionElement.Attributes("ref").Value
                                                Dim maxVerAttr = conditionElement.Attributes("maxVersionIndex")
                                                If maxVerAttr IsNot Nothing Then entry.MaxVersion = Integer.Parse(maxVerAttr.Value)
                                                Dim minVerAttr = conditionElement.Attributes("minVersionIndex")
                                                If minVerAttr IsNot Nothing Then entry.MinVersion = Integer.Parse(minVerAttr.Value)
                                                definition.Entries.Add(entry)
                                            End If
                                        Next
                                        Exit For
                                    End If
                                Next
                                definition.DefinedIn = admx
                                admx.SupportedOnDefinitions.Add(definition)
                            Next
                        ElseIf supportInfo.LocalName = "products" Then ' Product definitions
                            Dim loadProducts As Action(Of XmlNode, String, AdmxProduct)
                            loadProducts = Sub(Node As XmlNode, ChildTagName As String, Parent As AdmxProduct)
                                               For Each subproductElement As XmlNode In Node.ChildNodes
                                                   If subproductElement.LocalName <> ChildTagName Then Continue For
                                                   Dim product As New AdmxProduct
                                                   product.ID = subproductElement.Attributes("name").Value
                                                   product.DisplayCode = subproductElement.Attributes("displayName").Value
                                                   If Parent IsNot Nothing Then product.Version = subproductElement.Attributes("versionIndex").Value
                                                   product.Parent = Parent
                                                   product.DefinedIn = admx
                                                   admx.Products.Add(product)
                                                   If Parent Is Nothing Then
                                                       product.Type = AdmxProductType.Product
                                                       loadProducts(subproductElement, "majorVersion", product)
                                                   ElseIf Parent.Parent Is Nothing Then
                                                       product.Type = AdmxProductType.MajorRevision
                                                       loadProducts(subproductElement, "minorVersion", product)
                                                   Else
                                                       product.Type = AdmxProductType.MinorRevision
                                                   End If
                                               Next
                                           End Sub
                            loadProducts(supportInfo, "product", Nothing) ' Start the recursive load
                        End If
                    Next
                Case "categories" ' Categories
                    For Each categoryElement As XmlNode In child.ChildNodes
                        If categoryElement.LocalName <> "category" Then Continue For
                        Dim category As New AdmxCategory
                        category.ID = categoryElement.Attributes("name").Value
                        category.DisplayCode = categoryElement.Attributes("displayName").Value
                        category.ExplainCode = categoryElement.AttributeOrNull("explainText")
                        If categoryElement.HasChildNodes Then
                            Dim parentCatElement = categoryElement("parentCategory")
                            category.ParentID = parentCatElement.Attributes("ref").Value
                        End If
                        category.DefinedIn = admx
                        admx.Categories.Add(category)
                    Next
                Case "policies" ' Policy settings
                    Dim loadRegItem = Function(Node As XmlNode) As PolicyRegistryValue
                                          Dim regItem As New PolicyRegistryValue
                                          For Each subElement As XmlNode In Node.ChildNodes
                                              If subElement.LocalName = "delete" Then
                                                  regItem.RegistryType = PolicyRegistryValueType.Delete
                                                  Exit For
                                              ElseIf subElement.LocalName = "decimal" Then
                                                  regItem.RegistryType = PolicyRegistryValueType.Numeric
                                                  regItem.NumberValue = subElement.Attributes("value").Value
                                                  Exit For
                                              ElseIf subElement.LocalName = "string" Then
                                                  regItem.RegistryType = PolicyRegistryValueType.Text
                                                  regItem.StringValue = subElement.InnerText
                                                  Exit For
                                              End If
                                          Next
                                          Return regItem
                                      End Function
                    Dim loadOneRegList = Function(Node As XmlNode) As PolicyRegistrySingleList
                                             Dim singleList As New PolicyRegistrySingleList
                                             singleList.DefaultRegistryKey = AttributeOrNull(Node, "defaultKey")
                                             singleList.AffectedValues = New List(Of PolicyRegistryListEntry)
                                             For Each itemElement As XmlNode In Node.ChildNodes
                                                 If itemElement.LocalName <> "item" Then Continue For
                                                 Dim listEntry As New PolicyRegistryListEntry
                                                 listEntry.RegistryValue = itemElement.Attributes("valueName").Value
                                                 listEntry.RegistryKey = AttributeOrNull(itemElement, "key")
                                                 For Each valElement As XmlNode In itemElement.ChildNodes
                                                     If valElement.LocalName = "value" Then
                                                         listEntry.Value = loadRegItem(valElement)
                                                         Exit For
                                                     End If
                                                 Next
                                                 singleList.AffectedValues.Add(listEntry)
                                             Next
                                             Return singleList
                                         End Function
                    Dim loadOnOffValList = Function(OnValueName As String, OffValueName As String, OnListName As String, OffListName As String, Node As XmlNode) As PolicyRegistryList
                                               Dim regList As New PolicyRegistryList
                                               For Each subElement As XmlNode In Node.ChildNodes
                                                   If subElement.Name = OnValueName Then
                                                       regList.OnValue = loadRegItem(subElement)
                                                   ElseIf subElement.Name = OffValueName Then
                                                       regList.OffValue = loadRegItem(subElement)
                                                   ElseIf subElement.Name = OnListName Then
                                                       regList.OnValueList = loadOneRegList(subElement)
                                                   ElseIf subElement.Name = OffListName Then
                                                       regList.OffValueList = loadOneRegList(subElement)
                                                   End If
                                               Next
                                               Return regList
                                           End Function
                    For Each polElement As XmlNode In child.ChildNodes
                        If polElement.LocalName <> "policy" Then Continue For
                        Dim policy As New AdmxPolicy
                        policy.ID = polElement.Attributes("name").Value
                        policy.DefinedIn = admx
                        policy.DisplayCode = polElement.Attributes("displayName").Value
                        policy.RegistryKey = polElement.Attributes("key").Value
                        Dim polClass = polElement.Attributes("class").Value
                        Select Case polClass
                            Case "Machine"
                                policy.Section = AdmxPolicySection.Machine
                            Case "User"
                                policy.Section = AdmxPolicySection.User
                            Case Else
                                policy.Section = AdmxPolicySection.Both
                        End Select
                        policy.ExplainCode = AttributeOrNull(polElement, "explainText")
                        policy.PresentationID = AttributeOrNull(polElement, "presentation")
                        policy.ClientExtension = AttributeOrNull(polElement, "clientExtension")
                        policy.RegistryValue = AttributeOrNull(polElement, "valueName")
                        policy.AffectedValues = loadOnOffValList("enabledValue", "disabledValue", "enabledList", "disabledList", polElement)
                        For Each polInfo As XmlNode In polElement.ChildNodes
                            Select Case polInfo.LocalName
                                Case "parentCategory"
                                    policy.CategoryID = polInfo.Attributes("ref").Value
                                Case "supportedOn"
                                    policy.SupportedCode = polInfo.Attributes("ref").Value
                                Case "elements"
                                    policy.Elements = New List(Of PolicyElement)
                                    For Each uiElement As XmlNode In polInfo.ChildNodes
                                        Dim entry As PolicyElement = Nothing
                                        Select Case uiElement.LocalName
                                            Case "decimal"
                                                Dim decimalEntry As New DecimalPolicyElement
                                                decimalEntry.Minimum = AttributeOrDefault(uiElement, "minValue", 0)
                                                decimalEntry.Maximum = AttributeOrDefault(uiElement, "maxValue", UInteger.MaxValue)
                                                decimalEntry.NoOverwrite = AttributeOrDefault(uiElement, "soft", False)
                                                decimalEntry.StoreAsText = AttributeOrDefault(uiElement, "storeAsText", False)
                                                entry = decimalEntry
                                            Case "boolean"
                                                Dim boolEntry As New BooleanPolicyElement
                                                boolEntry.AffectedRegistry = loadOnOffValList("trueValue", "falseValue", "trueList", "falseList", uiElement)
                                                entry = boolEntry
                                            Case "text"
                                                Dim textEntry As New TextPolicyElement
                                                textEntry.MaxLength = AttributeOrDefault(uiElement, "maxLength", 255)
                                                textEntry.Required = AttributeOrDefault(uiElement, "required", False)
                                                textEntry.RegExpandSz = AttributeOrDefault(uiElement, "expandable", False)
                                                textEntry.NoOverwrite = AttributeOrDefault(uiElement, "soft", False)
                                                entry = textEntry
                                            Case "list"
                                                Dim listEntry As New ListPolicyElement
                                                listEntry.NoPurgeOthers = AttributeOrDefault(uiElement, "additive", False)
                                                listEntry.RegExpandSz = AttributeOrDefault(uiElement, "expandable", False)
                                                listEntry.UserProvidesNames = AttributeOrDefault(uiElement, "explicitValue", False)
                                                listEntry.HasPrefix = (uiElement.Attributes("valuePrefix") IsNot Nothing)
                                                listEntry.RegistryValue = AttributeOrNull(uiElement, "valuePrefix")
                                                entry = listEntry
                                            Case "enum"
                                                Dim enumEntry As New EnumPolicyElement
                                                enumEntry.Required = AttributeOrDefault(uiElement, "required", False)
                                                enumEntry.Items = New List(Of EnumPolicyElementItem)
                                                For Each itemElement As XmlNode In uiElement.ChildNodes
                                                    If itemElement.LocalName = "item" Then
                                                        Dim enumItem As New EnumPolicyElementItem
                                                        enumItem.DisplayCode = itemElement.Attributes("displayName").Value
                                                        For Each valElement As XmlNode In itemElement.ChildNodes
                                                            If valElement.LocalName = "value" Then
                                                                enumItem.Value = loadRegItem(valElement)
                                                            ElseIf valElement.LocalName = "valueList" Then
                                                                enumItem.ValueList = loadOneRegList(valElement)
                                                            End If
                                                        Next
                                                        enumEntry.Items.Add(enumItem)
                                                    End If
                                                Next
                                                entry = enumEntry
                                            Case "multiText"
                                                entry = New MultiTextPolicyElement
                                        End Select
                                        If entry IsNot Nothing Then
                                            entry.ClientExtension = AttributeOrNull(uiElement, "clientExtension")
                                            entry.RegistryKey = AttributeOrNull(uiElement, "key")
                                            If entry.RegistryValue = "" Then entry.RegistryValue = AttributeOrNull(uiElement, "valueName")
                                            entry.ID = uiElement.Attributes("id").Value
                                            entry.ElementType = uiElement.LocalName
                                            policy.Elements.Add(entry)
                                        End If
                                    Next
                            End Select
                        Next
                        admx.Policies.Add(policy)
                    Next
            End Select
        Next
        Return admx
    End Function
End Class

File: PolicyPlus/AdmxFile.cs
using System;
using System.Collections.Generic;
using System.Xml;
using Microsoft.VisualBasic.CompilerServices;

namespace PolicyPlus
{
    public class AdmxFile
    {
        public string SourceFile;
        public string AdmxNamespace;
        public string SupersededAdm;
        public decimal MinAdmlVersion;
        public Dictionary<string, string> Prefixes = new Dictionary<string, string>();
        public List<AdmxProduct> Products = new List<AdmxProduct>();
        public List<AdmxSupportDefinition> SupportedOnDefinitions = new List<AdmxSupportDefinition>();
        public List<AdmxCategory> Categories = new List<AdmxCategory>();
        public List<AdmxPolicy> Policies = new List<AdmxPolicy>();
        private AdmxFile()
        {
        }
        public static AdmxFile Load(string File)
        {
            // ADMX documentation: https://technet.microsoft.com/en-us/library/cc772138(v=ws.10).aspx
            var admx = new AdmxFile();
            admx.SourceFile = File;
            var xmlDoc = new XmlDocument();
            xmlDoc.Load(File);
            var policyDefinitions = xmlDoc.GetElementsByTagName("policyDefinitions")[0];
            foreach (XmlNode child in policyDefinitions.ChildNodes)
            {
                switch (child.LocalName ?? "")
                {
                    case "policyNamespaces": // Referenced namespaces and current namespace
                        {
                            foreach (XmlNode policyNamespace in child.ChildNodes)
                            {
                                string prefix = policyNamespace.Attributes["prefix"].Value;
                                string fqNamespace = policyNamespace.Attributes["namespace"].Value;
                                if (policyNamespace.LocalName == "target")
                                    admx.AdmxNamespace = fqNamespace;
                                admx.Prefixes.Add(prefix, fqNamespace);
                            }

                            break;
                        }
                    case "supersededAdm": // The ADM file that this ADMX supersedes
                        {
                            admx.SupersededAdm = child.Attributes["fileName"].Value;
                            break;
                        }
                    case "resources": // Minimum required version
                        {
                            admx.MinAdmlVersion = decimal.Parse(child.Attributes["minRequiredRevision"].Value, System.Globalization.CultureInfo.InvariantCulture);
                            break;
                        }
                    case "supportedOn": // Support definitions
                        {
                            foreach (XmlNode supportInfo in child.ChildNodes)
                            {
                                if (supportInfo.LocalName == "definitions")
                                {
                                    foreach (XmlNode supportDef in supportInfo.ChildNodes)
                                    {
                                        if (supportDef.LocalName != "definition")
                                            continue;
                                        var definition = new AdmxSupportDefinition();
                                        definition.ID = supportDef.Attributes["name"].Value;
                                        definition.DisplayCode = supportDef.Attributes["displayName"].Value;
                                        definition.Logic = AdmxSupportLogicType.Blank;
                                        foreach (XmlNode logicElement in supportDef.ChildNodes)
                                        {
                                            bool canLoad = true;
                                            if (logicElement.LocalName == "or")
                                            {
                                                definition.Logic = AdmxSupportLogicType.AnyOf;
                                            }
                                            else if (logicElement.LocalName == "and")
                                            {
                                                definition.Logic = AdmxSupportLogicType.AllOf;
                                            }
                                            else
                                            {
                                                canLoad = false;
                                            }
                                            if (canLoad)
                                            {
                                                definition.Entries = new List<AdmxSupportEntry>();
                                                foreach (XmlNode conditionElement in logicElement.ChildNodes)
                                                {
                                                    if (conditionElement.LocalName == "reference")
                                                    {
                                                        string product = conditionElement.Attributes["ref"].Value;
                                                        definition.Entries.Add(new AdmxSupportEntry() { ProductID = product, IsRange = false });
                                                    }
                                                    else if (conditionElement.LocalName == "range")
                                                    {
                                                        var entry = new AdmxSupportEntry() { IsRange = true };
                                                        entry.ProductID = conditionElement.Attributes["ref"].Value;
                                                        var maxVerAttr = conditionElement.Attributes["maxVersionIndex"];
                                                        if (maxVerAttr is not null)
                                                            entry.MaxVersion = int.Parse(maxVerAttr.Value);
                                                        var minVerAttr = conditionElement.Attributes["minVersionIndex"];
                                                        if (minVerAttr is not null)
                                                            entry.MinVersion = int.Parse(minVerAttr.Value);
                                                        definition.Entries.Add(entry);
                                                    }
                                                }
                                                break;
                                            }
                                        }
                                        definition.DefinedIn = admx;
                                        admx.SupportedOnDefinitions.Add(definition);
                                    }
                                }
                                else if (supportInfo.LocalName == "products") // Product definitions
                                {
                                    Action<XmlNode, string, AdmxProduct> loadProducts;
                                    loadProducts = new Action<XmlNode, string, AdmxProduct>((Node, ChildTagName, Parent) => { foreach (XmlNode subproductElement in Node.ChildNodes) { if ((subproductElement.LocalName ?? "") != (ChildTagName ?? "")) continue; var product = new AdmxProduct(); product.ID = subproductElement.Attributes["name"].Value; product.DisplayCode = subproductElement.Attributes["displayName"].Value; if (Parent is not null) product.Version = Conversions.ToInteger(subproductElement.Attributes["versionIndex"].Value); product.Parent = Parent; product.DefinedIn = admx; admx.Products.Add(product); if (Parent is null) { product.Type = AdmxProductType.Product; loadProducts(subproductElement, "majorVersion", product); } else if (Parent.Parent is null) { product.Type = AdmxProductType.MajorRevision; loadProducts(subproductElement, "minorVersion", product); } else { product.Type = AdmxProductType.MinorRevision; } } });
                                    loadProducts(supportInfo, "product", null); // Start the recursive load
                                }
                            }

                            break;
                        }
                    case "categories": // Categories
                        {
                            foreach (XmlNode categoryElement in child.ChildNodes)
                            {
                                if (categoryElement.LocalName != "category")
                                    continue;
                                var category = new AdmxCategory();
                                category.ID = categoryElement.Attributes["name"].Value;
                                category.DisplayCode = categoryElement.Attributes["displayName"].Value;
                                category.ExplainCode = categoryElement.AttributeOrNull("explainText");
                                if (categoryElement.HasChildNodes)
                                {
                                    var parentCatElement = categoryElement["parentCategory"];
                                    category.ParentID = parentCatElement.Attributes["ref"].Value;
                                }
                                category.DefinedIn = admx;
                                admx.Categories.Add(category);
                            }

                            break;
                        }
                    case "policies": // Policy settings
                        {
                            PolicyRegistryValue loadRegItem(XmlNode Node)
                            {
                                var regItem = new PolicyRegistryValue();
                                foreach (XmlNode subElement in Node.ChildNodes)
                                {
                                    if (subElement.LocalName == "delete")
                                    {
                                        regItem.RegistryType = PolicyRegistryValueType.Delete;
                                        break;
                                    }
                                    else if (subElement.LocalName == "decimal")
                                    {
                                        regItem.RegistryType = PolicyRegistryValueType.Numeric;
                                        regItem.NumberValue = Conversions.ToUInteger(subElement.Attributes["value"].Value);
                                        break;
                                    }
                                    else if (subElement.LocalName == "string")
                                    {
                                        regItem.RegistryType = PolicyRegistryValueType.Text;
                                        regItem.StringValue = subElement.InnerText;
                                        break;
                                    }
                                }
                                return regItem;
                            };
                            PolicyRegistrySingleList loadOneRegList(XmlNode Node)
                            {
                                var singleList = new PolicyRegistrySingleList();
                                singleList.DefaultRegistryKey = Node.AttributeOrNull("defaultKey");
                                singleList.AffectedValues = new List<PolicyRegistryListEntry>();
                                foreach (XmlNode itemElement in Node.ChildNodes)
                                {
                                    if (itemElement.LocalName != "item")
                                        continue;
                                    var listEntry = new PolicyRegistryListEntry();
                                    listEntry.RegistryValue = itemElement.Attributes["valueName"].Value;
                                    listEntry.RegistryKey = itemElement.AttributeOrNull("key");
                                    foreach (XmlNode valElement in itemElement.ChildNodes)
                                    {
                                        if (valElement.LocalName == "value")
                                        {
                                            listEntry.Value = loadRegItem(valElement);
                                            break;
                                        }
                                    }
                                    singleList.AffectedValues.Add(listEntry);
                                }
                                return singleList;
                            };
                            PolicyRegistryList loadOnOffValList(string OnValueName, string OffValueName, string OnListName, string OffListName, XmlNode Node)
                            {
                                var regList = new PolicyRegistryList();
                                foreach (XmlNode subElement in Node.ChildNodes)
                                {
                                    if ((subElement.Name ?? "") == (OnValueName ?? ""))
                                    {
                                        regList.OnValue = loadRegItem(subElement);
                                    }
                                    else if ((subElement.Name ?? "") == (OffValueName ?? ""))
                                    {
                                        regList.OffValue = loadRegItem(subElement);
                                    }
                                    else if ((subElement.Name ?? "") == (OnListName ?? ""))
                                    {
                                        regList.OnValueList = loadOneRegList(subElement);
                                    }
                                    else if ((subElement.Name ?? "") == (OffListName ?? ""))
                                    {
                                        regList.OffValueList = loadOneRegList(subElement);
                                    }
                                }
                                return regList;
                            };
                            foreach (XmlNode polElement in child.ChildNodes)
                            {
                                if (polElement.LocalName != "policy")
                                    continue;
                                var policy = new AdmxPolicy();
                                policy.ID = polElement.Attributes["name"].Value;
                                policy.DefinedIn = admx;
                                policy.DisplayCode = polElement.Attributes["displayName"].Value;
                                policy.RegistryKey = polElement.Attributes["key"].Value;
                                string polClass = polElement.Attributes["class"].Value;
                                switch (polClass ?? "")
                                {
                                    case "Machine":
                                        {
                                            policy.Section = AdmxPolicySection.Machine;
                                            break;
                                        }
                                    case "User":
                                        {
                                            policy.Section = AdmxPolicySection.User;
                                            break;
                                        }

                                    default:
                                        {
                                            policy.Section = AdmxPolicySection.Both;
                                            break;
                                        }
                                }
                                policy.ExplainCode = polElement.AttributeOrNull("explainText");
                                policy.PresentationID = polElement.AttributeOrNull("presentation");
                                policy.ClientExtension = polElement.AttributeOrNull("clientExtension");
                                policy.RegistryValue = polElement.AttributeOrNull("valueName");
                                policy.AffectedValues = loadOnOffValList("enabledValue", "disabledValue", "enabledList", "disabledList", polElement);
                                foreach (XmlNode polInfo in polElement.ChildNodes)
                                {
                                    switch (polInfo.LocalName ?? "")
                                    {
                                        case "parentCategory":
                                            {
                                                policy.CategoryID = polInfo.Attributes["ref"].Value;
                                                break;
                                            }
                                        case "supportedOn":
                                            {
                                                policy.SupportedCode = polInfo.Attributes["ref"].Value;
                                                break;
                                            }
                                        case "elements":
                                            {
                                                policy.Elements = new List<PolicyElement>();
                                                foreach (XmlNode uiElement in polInfo.ChildNodes)
                                                {
                                                    PolicyElement entry = null;
                                                    switch (uiElement.LocalName ?? "")
                                                    {
                                                        case "decimal":
                                                            {
                                                                var decimalEntry = new DecimalPolicyElement();
                                                                decimalEntry.Minimum = Conversions.ToUInteger(uiElement.AttributeOrDefault("minValue", 0));
                                                                decimalEntry.Maximum = Conversions.ToUInteger(uiElement.AttributeOrDefault("maxValue", uint.MaxValue));
                                                                decimalEntry.NoOverwrite = Conversions.ToBoolean(uiElement.AttributeOrDefault("soft", false));
                                                                decimalEntry.StoreAsText = Conversions.ToBoolean(uiElement.AttributeOrDefault("storeAsText", false));
                                                                entry = decimalEntry;
                                                                break;
                                                            }
                                                        case "boolean":
                                                            {
                                                                var boolEntry = new BooleanPolicyElement();
                                                                boolEntry.AffectedRegistry = loadOnOffValList("trueValue", "falseValue", "trueList", "falseList", uiElement);
                                                                entry = boolEntry;
                                                                break;
                                                            }
                                                        case "text":
                                                            {
                                                                var textEntry = new TextPolicyElement();
                                                                textEntry.MaxLength = Conversions.ToInteger(uiElement.AttributeOrDefault("maxLength", 255));
                                                                textEntry.Required = Conversions.ToBoolean(uiElement.AttributeOrDefault("required", false));
                                                                textEntry.RegExpandSz = Conversions.ToBoolean(uiElement.AttributeOrDefault("expandable", false));
                                                                textEntry.NoOverwrite = Conversions.ToBoolean(uiElement.AttributeOrDefault("soft", false));
                                                                entry = textEntry;
                                                                break;
                                                            }
                                                        case "list":
                                                            {
                                                                var listEntry = new ListPolicyElement();
                                                                listEntry.NoPurgeOthers = Conversions.ToBoolean(uiElement.AttributeOrDefault("additive", false));
                                                                listEntry.RegExpandSz = Conversions.ToBoolean(uiElement.AttributeOrDefault("expandable", false));
                                                                listEntry.UserProvidesNames = Conversions.ToBoolean(uiElement.AttributeOrDefault("explicitValue", false));
                                                                listEntry.HasPrefix = uiElement.Attributes["valuePrefix"] is not null;
                                                                listEntry.RegistryValue = uiElement.AttributeOrNull("valuePrefix");
                                                                entry = listEntry;
                                                                break;
                                                            }
                                                        case "enum":
                                                            {
                                                                var enumEntry = new EnumPolicyElement();
                                                                enumEntry.Required = Conversions.ToBoolean(uiElement.AttributeOrDefault("required", false));
                                                                enumEntry.Items = new List<EnumPolicyElementItem>();
                                                                foreach (XmlNode itemElement in uiElement.ChildNodes)
                                                                {
                                                                    if (itemElement.LocalName == "item")
                                                                    {
                                                                        var enumItem = new EnumPolicyElementItem();
                                                                        enumItem.DisplayCode = itemElement.Attributes["displayName"].Value;
                                                                        foreach (XmlNode valElement in itemElement.ChildNodes)
                                                                        {
                                                                            if (valElement.LocalName == "value")
                                                                            {
                                                                                enumItem.Value = loadRegItem(valElement);
                                                                            }
                                                                            else if (valElement.LocalName == "valueList")
                                                                            {
                                                                                enumItem.ValueList = loadOneRegList(valElement);
                                                                            }
                                                                        }
                                                                        enumEntry.Items.Add(enumItem);
                                                                    }
                                                                }
                                                                entry = enumEntry;
                                                                break;
                                                            }
                                                        case "multiText":
                                                            {
                                                                entry = new MultiTextPolicyElement();
                                                                break;
                                                            }
                                                    }
                                                    if (entry is not null)
                                                    {
                                                        entry.ClientExtension = uiElement.AttributeOrNull("clientExtension");
                                                        entry.RegistryKey = uiElement.AttributeOrNull("key");
                                                        if (string.IsNullOrEmpty(entry.RegistryValue))
                                                            entry.RegistryValue = uiElement.AttributeOrNull("valueName");
                                                        entry.ID = uiElement.Attributes["id"].Value;
                                                        entry.ElementType = uiElement.LocalName;
                                                        policy.Elements.Add(entry);
                                                    }
                                                }

                                                break;
                                            }
                                    }
                                }
                                admx.Policies.Add(policy);
                            }

                            break;
                        }
                }
            }
            return admx;
        }
    }
}